src/fu-engine.c

```diff
@@ -3031,8 +3031,10 @@ fu_engine_prepare(FuEngine *self,
 	    !fu_device_has_flag(device, FWUPD_DEVICE_FLAG_EMULATED)) {
 		if (!fu_engine_backends_save_phase(self, error))
 			return FALSE;
-		if (!fu_engine_backends_clear_phase(self, error))
-			return FALSE;
+		if (fu_device_get_composite_id(device) == NULL) {
```

Comment:
The problem for composite updates is that you can't clear the USB events for a given device in between stages because a bunch of devices may use that USB device.  With 52790d8 reverted, you end up with:

```
Failed to query dock info: write over HID-I2C failed: failed after 5 retries: failed to SetReport: no matching event for ControlTransfer:Direction=0x01,RequestType=0x01,Recipient=0x01,Request=0x09,Value=0x0200,Idx=0x0000,Data=QMYAAAAAGgDsAIAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEYAQEABwAFBwQAAAEiAAABVwAAYAABADYBAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA,Length=0xc0
```

Reply:
I guess another way to look at it is that there ia mismatch in the engine for the concept of "install phase" for composite device.  A given _device_ goes through all the install phases, it's not a full engine state.

* `fu_engine_install_releases`: Iterates over releases/devices.  Calls `fu_engine_install_release`
* `fu_engine_install_release`: Goes over one release/device.  Calls `fu_engine_install_blob`
* `fu_engine_install_blob`: Goes over one device w/ a blob.  Calls `fu_engine_backends_save_phase` a bunch of times for each install phase.